docs/proposals/hpa-custom-metrics.md

```diff
@@ -0,0 +1,162 @@
+(e.g. memory usage).  While the current design does allow the user to specify
+a metric name, all metrics are prefixed with "custom/", meaning built-in
+metrics cannot be passed this way.  A new design should not have completely
+separate mechanisms for custom and non-custom metrics.  This also allows
+refering to raw CPU usage rate, instead of refering to CPU usage as a
+percentage of the CPU request (this has been a commonly asked-for feature
+by users and operators).
+
+### Refering to Metrics Not Associated With Pods ###
+
+Heapster current can track metrics associated with pods and namespaces.  With
+Heapster push metrics, it becomes possible to define custom metrics on the
+namespace level.  However, the custom metrics annotation in the HPA currently
+only allows referring to custom metrics defined on pods.
+
+Refering to custom metrics at the namespace level allows us to support certain
+use cases in which an application wishes to expose a certain special metric
+to an entire namespace (e.g. queue length) and scale based on that.
+Additionally, until Heapster has the capability to track metrics from service
+objects, RCs, etc, namespace-level metrics can be used with push metrics to
+"fake" this type of metrics.  Metrics associated with services could be
+extracted from a load balancer or reverse proxy, for instance.
+
+Heapster will, most likely, support metrics associated with services,
+replicationcontrollers, etc, in the future, so it becomes adventageous to be
+able to refer to metrics associated with those objects as well.
+
+Proposed New Design
+-------------------
+
+```go
+type SourceType string
+var (
+    SourceTypeNamespace SourceType = "namespace"  // the current namespace
+    SourceTypePod SourceType = "pod"  // each pod that would be considered
+    SourceTypeController SourceType = "controller" // the target of the HPA
+)
+
+type MetricSource struct {
+    // Draw from the current namespace, pod, or controller
+    CurrentSource SourceType `json:"current,omitempty`
+
+    // Draw from an object in the current namespace (not currently supported
+    // by Heapster)
+    SourceRef *CrossVersionObjectReference `json:"object,omitempty"`
+}
+
+type MetricTarget struct {
+    // users could either specify
+    // `custom: my-metric` or `name: custom/my-metric`
+    // builtin metrics would be specified with `name: cpu/usage`
+    CustomName string `json:"custom,omitempty"`
+    BuiltinName string `json:"name,omitempty"`
+
+    TargetValue resource.Quantity `json:"targetValue"`
+
+    // Specifies the target object
+    // (missing is equivalent to {"current": "pod"})
+    SourceObject *MetricSource `json:"from,omitempty"`
+}
+
+type HorizontalPodAutoscalerSpec struct {
+    ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef"`
+    MinReplicas *int32 `json:"minReplicas,omitempty"`
+    MaxReplicas int32 `json:"maxReplicas"`
+    Metrics []MetricTarget `json:"metrics"`
```

Comment:
Initially, I'd imagine it would be the same as what happens now (which is that the maximum is used), although I can see arguments for other behaviors.  Perhaps it should be configurable (min, max, average)?